Add support for reduced motion/disable animations on the web

ditman opened the pull request:
This PR listens to changes to the value of the `(prefers-reduced-motion: reduce)` media query to update the `reduceMotion` and `disableAnimations` properties of the `AccessibilityFeatures` configuration object.

This is achieved by introducing a new `MediaQueryManager` object to the `EnginePlatformDispatcher` that handles handles all the engine configuration values that are driven by media queries: "reduced motion", "high contrast" and "dark mode".

Unifying the code paths of all these values, allows this PR to slightly increase testing coverage of those values, by injecting fake browser events in the `platform_dispatcher_test.dart` file.

### Issues

* Fixes #167566

### Testing

* Added browser tests for new (and existing) media-queries
* Deployed test app to: https://dit-tests.web.app (may get offline!)

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

**Note**: The Flutter team is currently trialing the use of [Gemini Code Assist for GitHub](https://developers.google.com/gemini-code-assist/docs/review-github-code). Comments from the `gemini-code-assist` bot should not be taken as authoritative feedback from the Flutter team. If you find its comments useful you can update your code accordingly, but if you are unsure or disagree with the feedback, please feel free to wait for a Flutter team member's review for guidance on which automated comments should be addressed.

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

ditman committed:
Listen to prefers-reduced-motion mediaquery and update accessibility features with its value.

ditman committed:
dart format

mdebbar commented on engine/src/flutter/lib/web_ui/lib/src/engine/platform_dispatcher.dart:
```diff
@@ -1243,33 +1247,37 @@ class EnginePlatformDispatcher extends ui.PlatformDispatcher {
+      onMatch: (prefersDark) {
+        _updatePlatformBrightness(prefersDark ? ui.Brightness.dark : ui.Brightness.light);
+      },
+    );
```
nit: Should we change `_updatePlatformBrightness` to take `bool prefersDark` as parameter? That would make the `onMatch` callback look cleaner:

```suggestion
    _mediaQueries.addListener('(prefers-color-scheme: dark)', onMatch: _updatePlatformBrightness);
```

mdebbar reviewed (approved):
All my comments are minor suggestions that can be ignored, but Gemini's comment deserves to be addressed IMO.

Thanks for the clean MediaQueryManager idea!

ditman commented on engine/src/flutter/lib/web_ui/lib/src/engine/platform_dispatcher/media_query_listeners.dart:
```diff
@@ -0,0 +1,104 @@
+
+    // Call onMatch with the initial value
+    onMatch(_getMatchesValue(mediaQuery));
+  }
```
I used a slightly different pattern than the suggested one, but TL;DR, I'll only create the mediaQuery object if we need to create a `_MediaQueryListeners` object.

ditman committed:
Address first round of PR comments.

ditman committed:
dart format

ditman commented on engine/src/flutter/lib/web_ui/lib/src/engine/platform_dispatcher.dart:
```diff
@@ -1233,43 +1237,44 @@ class EnginePlatformDispatcher extends ui.PlatformDispatcher {
+        accessibilityFeatures: original.copyWith(highContrast: enabled),
       );
       invokeOnPlatformConfigurationChanged();
+      invokeOnAccessibilityFeaturesChanged();
```
This call to `invokeOnAccessibilityFeaturesChanged()` fixes a small unreported bug, by which the `MediaQuery.highContrastOf(context);` wouldn't update in real time, and instead would require an app reload to pick up the new value.

ditman committed:
Rename media_query_listeners to media_query_manager to match the main class on the file

ditman committed:
Add missing license header

ditman commented on engine/src/flutter/lib/web_ui/lib/src/engine/platform_dispatcher/media_query_manager.dart:
```diff
@@ -0,0 +1,142 @@
+  static const FORCED_COLORS = '(forced-colors: active)';
+
+  /// Detects reduced motion.
+  static const REDUCED_MOTION = '(prefers-reduced-motion: reduce)';
```
These could be a specific enum, instead of consts in the manager class.

ditman commented on engine/src/flutter/lib/web_ui/lib/src/engine/platform_dispatcher/media_query_manager.dart:
```diff
@@ -0,0 +1,142 @@
+    final _MediaQueryListeners listeners = _listeners.putIfAbsent(mediaQueryString, () {
+      // Create a proper media query object
+      final DomEventTarget mediaQuery = _createMediaQuery(mediaQueryString);
+      return _MediaQueryListeners(mediaQuery);
```
These could be a single line:

```dart
.putIfAbsent(mQS, () => _MediaQueryListeners(_createMediaQuery(...))
```

ditman commented on engine/src/flutter/lib/web_ui/lib/src/engine/high_contrast.dart:
```diff
@@ -1,53 +1,25 @@
+/// This used to do a bunch of stuff with media queries, but now all of this
+/// is handled by the [MediaQueryManager] object of the
+/// [EnginePlatformDispatcher]. See [_registerMediaQueryListeners].
 class HighContrastSupport {
```
Should this class completely be replaced by a static bool property in the platform dispatcher, instead (now that it doesn't do much?)

mdebbar replied:
I agree, or just remove it if the boolean isn't used anywhere?

ditman reviewed (commented):
PTAL @mdebbar!

mdebbar reviewed (approved):
![LGTM](https://user-images.githubusercontent.com/1278212/240958742-5849e437-2e5f-4b46-ae0a-254395368201.png)

ditman committed:
Remove HighContrastSupport singleton, and move SystemColorPaletteDetector to a more appropriate location.

ditman commented:
I removed the `HighContrastSupport` singleton, and moved the `high_contrast.dart` file to `platform_dispatcher/system_color_palette_detector.dart` that more accurately reflects what's going on now.
